File: BST/CPP/BSTV2.cpp
//
// Created by Shahriar Nasim Nafi on 8/7/22.
//

#include <iostream>

using namespace std;

struct Node {
    int value;
    Node *parent;
    Node *left;
    Node *right;

    void addLeft(Node *node) {
        node->parent = this;
        this->left = node;
    }

    void addRight(Node *node) {
        node->parent = this;
        this->right = node;
    }
};

struct BST {
public:
    Node *root = nullptr;

    Node *insert(int value) {
        Node *node = new Node();
        node->value = value;

        if (!root) {
            root = node;
            return node;
        }

        Node *parent = nullptr;
        Node *current = root;

        while (current) {
            parent = current;
            if (node->value > current->value) {
                current = current->right;
            } else {
                current = current->left;
            }
        }

        if (node->value > parent->value) {
            parent->addRight(node);
        } else {
            parent->addLeft(node);
        }

        return node;
    }

    Node *transplant(Node *current, Node *node) {
        if (root == current) {
            root = current;
            return root;
        }

        if (current->parent->left == current) {
            current->parent->addLeft(node);
        } else {
            current->parent->addRight(node);
        }
        return root;
    }

    Node *succesorNode(Node *node) {
        Node *current = node->right;

        Node *smallestNode;

        while (current->left) {
            current = current->left;
        }

        smallestNode = current;
        return smallestNode;
    }

    Node *deleteNode(Node *node) {
        Node *smallest;

        if (node->left == nullptr) {
            transplant(node, node->right);
        } else if (node->right == nullptr) {
            transplant(node, node->left);
        } else {
            smallest = succesorNode(node->right);

            if (smallest->parent != node) {
                root = transplant(smallest, smallest->right);
                smallest->addRight(node->right);

            }
            root = transplant(node, smallest);
            smallest->addLeft(node->left);

        }

        delete node;
        return root;
    }

    Node *d(Node *node) {
        Node *smallest;

        if (node->left == nullptr) {
            transplant(node, node->right);
        } else if (node->right == nullptr) {
            transplant(node, node->left);
        } else {
            smallest = succesorNode(node);
            if (smallest->parent != node) {
                root = transplant(smallest, smallest->right);
                smallest->addRight(node->right);

            }

            root = transplant(node, smallest);
            smallest->addLeft(node->left);
        }
        return root;
    }

    void print(Node *node, int rootValue) {
        if (node == nullptr) {
            return;
        }

        if (node->value == rootValue) {
            cout << rootValue << endl;
        } else {
            cout << rootValue << " -> " << node->value << endl;
        }

        print(node->left, node->value);
        print(node->right, node->value);
    }
};

File: BST/Swift/BSTV2.swift
//
// Created by Shahriar Nasim Nafi on 8/7/22.
//

struct BST<T: Comparable> {
    
    private(set) var root: Node?
    
    mutating func insert(value: T) -> Node {
        let node = Node(value: value)
        
        if root == nil {
            root = node
            return node
        }
        
        var current = root
        var parent = current
        
        while current != nil {
            parent = current
            if current!.value > value {
                current = current?.left
            } else {
                current = current?.right
            }
        }
        
        if parent!.value > value {
            parent?.addLeft(node: node)
        } else {
            parent?.addRight(node: node)
        }
        
        
        return node
    }
    
    func succesor(for node: Node) -> Node? {
        var current = node.right
        
        while(current?.left != nil) {
            current = current?.left
        }
        
        return current
    }
    
    mutating func transplant(node: Node, new: Node?) -> Node {
        if root === node {
            root = new
            return root!
        }
        
        if node.parent?.left === node {
            node.parent?.addLeft(node: new)
        } else if node.parent?.right === node {
            node.parent?.addRight(node: new)
        }
        
        return root!
        
    }
    
    mutating func remove(node: Node) {
       
        if let right = node.right, node.left == nil {
            transplant(node: node, new: right)
        } else if let left = node.left, node.right == nil {
            transplant(node: node, new: left)
        } else {
            if let smallest = succesor(for: node) {
                if smallest.parent !== node {
                    root = transplant(node: smallest, new: smallest.right)
                    smallest.addRight(node: node.right)
                }
                
                root = transplant(node: node, new: smallest)
                smallest.addLeft(node: node.left)
            }
        }
    }
    
    class Node {
        let value: T
        var parent: Node?
        var left: Node?
        var right: Node?
        
        init(value: T) {
            self.value = value
        }
        
        func addLeft(value: T) -> Node {
            let node = Node(value: value)
            node.parent = self
            self.left = node
            return node
        }
        
        func addRight(value: T) -> Node {
            let node = Node(value: value)
            node.parent = self
            self.right = node
            return node
        }
        
        func addLeft(node: Node?) -> Node? {
            node?.parent = self
            self.left = node
            return node
        }
        
        func addRight(node: Node?) -> Node? {
            node?.parent = self
            self.right = node
            return node
        }
    }
}
